aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+  bool defined() const noexcept {
+    return target_ != NullType::singleton();
+  }
+
+  size_t use_count() const noexcept {
+    if (target_ == NullType::singleton()) {
+      return 0;
+    }
+    return target_->refcount_.load();
+  }
+
+  bool unique() const noexcept {
+    return use_count() == 1;
+  }
+
+  template <class... Args>
+  static intrusive_ptr make(Args&&... args) {
+    auto result = intrusive_ptr(new TTarget(std::forward<Args>(args)...));
+    // We can't use retain(), because we also have to increase weakcount
+    // and because we allow raising these values from 0, which retain()
+    // has an assertion against.
+    ++result.target_->refcount_;
+    ++result.target_->weakcount_;
+
+    return result;
+  }
+};
+
+template <
+    class TTarget,
+    class NullType = detail::intrusive_target_default_null_type<TTarget>,
+    class... Args>
+inline intrusive_ptr<TTarget, NullType> make_intrusive(Args&&... args) {
+  return intrusive_ptr<TTarget, NullType>::make(std::forward<Args>(args)...);
+}
+
+template <class TTarget, class NullType>
+inline void swap(
+    intrusive_ptr<TTarget, NullType>& lhs,
+    intrusive_ptr<TTarget, NullType>& rhs) noexcept {
+  lhs.swap(rhs);
+}
+
+// To allow intrusive_ptr inside std::map or std::set, we need operator<
+template <class TTarget1, class NullType1, class TTarget2, class NullType2>
+inline bool operator<(
+    const intrusive_ptr<TTarget1, NullType1>& lhs,
+    const intrusive_ptr<TTarget2, NullType2>& rhs) noexcept {
+  return lhs.get() < rhs.get();
+}
+
+template <class TTarget1, class NullType1, class TTarget2, class NullType2>
+inline bool operator==(
+    const intrusive_ptr<TTarget1, NullType1>& lhs,
+    const intrusive_ptr<TTarget2, NullType2>& rhs) noexcept {
+  return lhs.get() == rhs.get();
+}
+
+template <class TTarget1, class NullType1, class TTarget2, class NullType2>
+inline bool operator!=(
+    const intrusive_ptr<TTarget1, NullType1>& lhs,
+    const intrusive_ptr<TTarget2, NullType2>& rhs) noexcept {
+  return !operator==(lhs, rhs);
+}
+
+template <
+    typename TTarget,
+    class NullType = detail::intrusive_target_default_null_type<TTarget>>
+class weak_intrusive_ptr final {
+ private:
+  static_assert(
+      std::is_base_of<intrusive_ptr_target, TTarget>::value,
+      "intrusive_ptr can only be used for classes that inherit from intrusive_ptr_target.");
+  static_assert(
+      NullType::singleton() == NullType::singleton(),
+      "NullType must have a constexpr singleton() method");
+  static_assert(
+      std::is_same<TTarget*, decltype(NullType::singleton())>::value,
+      "NullType::singleton() must return a element_type* pointer");
+
+  TTarget* target_;
+
+  template <class TTarget2, class NullType2>
+  friend class weak_intrusive_ptr;
+
+  void retain() noexcept {
+    if (target_ != NullType::singleton()) {
+      size_t new_weakcount = ++target_->weakcount_;
+      AT_CHECK(
+          new_weakcount != 1,
```

Comment:
AT_ASSERT.

Also, this one seems a bit goofy, because if weakcount really was zero before, then probably this memory had already been freed, and you're already in memory corruption land.

Reply:
True. It isn't foolproof because the memory might be overwritten. However, if this assert fires, we always know something's wrong.